src/docfx/lib/js/ChakraCoreJsEngine.cs

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using ChakraHost.Hosting;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    /// <summary>
+    /// Javascript engine based on https://github.com/microsoft/ChakraCore
+    /// </summary>
+    internal class ChakraCoreJsEngine : IJavaScriptEngine
+    {
+        // A pool of ChakraCore runtime and context. Create one context for each runtime.
+        private static readonly ConcurrentBag<JavaScriptContext> s_contextPool = new ConcurrentBag<JavaScriptContext>();
+
+        // Limit the maximum ChakraCore runtimes to current processor count.
+        private static readonly SemaphoreSlim s_contextThrottler = new SemaphoreSlim(Environment.ProcessorCount, Environment.ProcessorCount);
+
+        private static readonly ConcurrentDictionary<(JavaScriptContext, string scriptPath), JavaScriptValue> s_scriptExports
+                          = new ConcurrentDictionary<(JavaScriptContext, string scriptPath), JavaScriptValue>();
+
+        private static readonly JavaScriptNativeFunction s_requireFunction = new JavaScriptNativeFunction(Require);
+
+        private static readonly ThreadLocal<Stack<string>> t_dirnames = new ThreadLocal<Stack<string>>(() => new Stack<string>());
+
+        private static readonly ThreadLocal<Dictionary<string, JavaScriptValue>> t_modules = new ThreadLocal<Dictionary<string, JavaScriptValue>>();
+
+        private static JavaScriptSourceContext s_currentSourceContext = JavaScriptSourceContext.FromIntPtr(IntPtr.Zero);
+
+        private readonly ConcurrentDictionary<JavaScriptContext, JavaScriptValue> _globals = new ConcurrentDictionary<JavaScriptContext, JavaScriptValue>();
```

Comment:
should this be `ConcurrentDictionary<JavaScriptContext, Lazy<JavaScriptValue>>`? same as `s_scriptExports`

Reply:
evaluation of  `s_scriptExports` is relatively expensive, requires the script to run to end, so I put a lazy there, but the evaluation of `_globals` is very fast, just marshaling data.